ghpages.mk

```diff
@@ -84,6 +99,22 @@ ifeq (true,$(PUSH_GHPAGES))
 ifneq (,$(TARGET_DIR))
 	mkdir -p $(GHPAGES_TMP)/$(TARGET_DIR)
 endif
+
+	@EXCESS_FILES="$(addprefix $(TARGET_DIR),$(filter-out $^,$(notdir $(foreach ext,*.html *.txt,$(wildcard $(GHPAGES_TMP)/$(TARGET_DIR)/$(ext))))))"; \
+	if [ -n "$$EXCESS_FILES" ]; \
+		then git -C $(GHPAGES_TMP) rm -f --ignore-unmatch -- $$EXCESS_FILES; \
+	fi
+
+ifneq (0,$(words $(OBSOLETE_DIRECTORIES)))
```

Comment:
This is nice from an execution performance standpoint, but I think that it would be better to avoid the work of constructing the OBSOLETE_DIRECTORIES variable unless you are running this target.  If you take the above into account, this gets easier.  In shell (which needs more quoting):

```make
@cutoff=$$(date +%s -d '-30 days'); \
comm -13 <(git branch -a | sed -e 's,.*[ /],,' | sort | uniq) <(ls $(GHPAGES_TMP) | sed -e 's,.*/,,')) | while read obsolete; do \
  [ ! -d "$$obsolete" ] || \
    [ "$$cutoff" -le "$$(git -C $(GHPAGES_TMP) log -n 1 --format=%ct "$$obsolete")" ] || \
    echo "Remove obsolete '$$obsolete'" && git rm -rfq -- "$$obsolete"; \
done
```

You can then run the cleanup always.  The only trick would be to move the command that makes the target directory after this cleanup in case the directory is removed in this cleanup.  That is, the order is:
1. obsolete directory cleanup
2. mkdir $(TARGET_DIR)
3. cleanup $(TARGET_DIR)
4. add updated files
5. commit
6. push